postgres/lib/func.js

```diff
@@ -13,14 +13,18 @@ const StandardFunctions = {
   endswith: (x, y) => `substr(${x},length(${x}) + 1 - length(${y})) = ${y}`,
 
   // Date and Time Functions
-  year: x => `date_part('year',(${x})::TIMESTAMP)`,
-  month: x => `date_part('month',(${x})::TIMESTAMP)`,
-  day: x => `date_part('day',(${x})::TIMESTAMP)`,
-  hour: x => `date_part('hour',(${x})::TIMESTAMP)`,
-  minute: x => `date_part('minute',(${x})::TIMESTAMP)`,
-  second: x => `date_part('second',(${x})::TIMESTAMP)`,
+  year: x => `date_part('year', ${castVal(x)})`,
+  month: x => `date_part('month', ${castVal(x)})`,
+  day: x => `date_part('day', ${castVal(x)})`,
+  hour: x => `date_part('hour', ${castVal(x)})`,
+  minute: x => `date_part('minute', ${castVal(x)})`,
+  second: x => `date_part('second', ${castVal(x)})`,
 }
 
+const isTime = /^\d{1,2}:\d{1,2}:\d{1,2}$/
+const isVal = x => x && 'val' in x
+const castVal = (x) => `${x}${isVal(x) ? isTime.test(x.val) ? '::TIME' : '::TIMESTAMP' : ''}`
```

Comment:
Is there a reason to only cast vals? hour(deliveryDate) is also allowed, isn't it?

Reply:
Only `val` can have no types. `hour(deliveryDate)` will have a type already.